transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -0,0 +1,361 @@
+        } else {
+            acquireTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(acquireTimeoutMillis);
+            switch (action) {
+            case Fail:
+                timeoutTask = new TimeoutTask() {
+                    @Override
+                    public void onTimeout(AcquireTask task) {
+                        // Fail teh promise as we timed out.
+                        task.promise.setFailure(TIMEOUT_EXCEPTION);
+                    }
+                };
+                break;
+            case NewConnection:
+                timeoutTask = new TimeoutTask() {
+                    @Override
+                    public void onTimeout(AcquireTask task) {
+                        // Increment the acquire count and delegate to super to actually acquire a Channel which will
+                        // create a new connetion.
+                        ++acquiredChannelCount;
+
+                        FixedChannelPool.super.acquire(task.key, task.promise);
+                    }
+                };
+                break;
+            default:
+                throw new Error();
+            }
+        }
+        executor = bootstrap.group().next();
+        this.maxConnections = maxConnections;
+        this.maxPendingAcquires = maxPendingAcquires;
+    }
+
+    @Override
+    public Future<C> acquire(final K key, final Promise<C> promise) {
+        try {
+            if (executor.inEventLoop()) {
+                acquire0(key, promise);
+            } else {
+                executor.execute(new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        acquire0(key, promise);
+                    }
+                });
+            }
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+        return promise;
+    }
+
+    private void acquire0(K key, final Promise<C> promise) {
+        assert executor.inEventLoop();
+
+        if (acquiredChannelCount < maxConnections) {
+            ++acquiredChannelCount;
+
+            assert acquiredChannelCount > 0;
+
+            // We need to create a new promise as we need to ensure the AcquireListener runs in the correct
+            // EventLoop
+            Promise<C> p = executor.newPromise();
+            p.addListener(new AcquireListener(promise));
+            super.acquire(key, p);
+        } else {
+            if (pendingAcquireCount >= maxPendingAcquires) {
+                promise.setFailure(FULL_EXCEPTION);
+            } else {
+                ++pendingAcquireCount;
+                final ScheduledFuture<?> timeoutFuture;
+                if (timeoutTask != null) {
+                    timeoutFuture = executor.schedule(timeoutTask, acquireTimeoutNanos, TimeUnit.NANOSECONDS);
+                } else {
+                    timeoutFuture = null;
+                }
+
+                pendingAcquireQueue.offer(new AcquireTask(key, promise, timeoutFuture));
```

Comment:
We should check the return value of this call and react accordingly.  For example we should only do `++pendingAcquireCount` and schedule the timeout if this succeed, and fail the promise otherwise.